File: Projeto/Agenda.py
# Lista que armazena os contatos, cada contato é uma lista [nome, celular, email]
agenda = []

def menu():
    """
    Exibe o menu principal com opções disponíveis
    e solicita uma escolha válida do usuário.
    Retorna um inteiro entre 1 e 6.
    """
    print("""
    Escolha uma opção válida:
    1 - Novo Contato.
    2 - Editar Contato.
    3 - Pesquisar Contato.
    4 - Listar Contato.
    5 - Apagar Contato.
    6 - Sair.
    """)
    return validar("Escolha sua opção: ", 1, 6)

def validar(pergunta, inicio, fim):
    """
    Valida a entrada do usuário para garantir que seja um número inteiro dentro do intervalo especificado.
    Retorna o valor validado ou 0 se fora do intervalo.
    """
    while True:
        try:
            valor = int(input(pergunta))
            if inicio <= valor <= fim:
                return valor
            else:
                print(f"Por favor, digite um número entre {inicio} e {fim}.")
        except ValueError:
            print("Entrada inválida. Digite um número inteiro válido.")

def novo():
    """
    Solicita os dados de um novo contato e adiciona à agenda.
    """
    global agenda
    nome = p_nome()
    celular = input("Celular...: ")
    email = input("E-mail...: ")
    agenda.append([nome, celular, email])
    print("\n---------------------------")
    print("Contato salvo com sucesso!")
    print(f"Nome: {nome} | Celular: {celular} | E-mail: {email}")
    print("---------------------------\n")

def p_nome():
    """
    Solicita o nome do contato.
    """
    return input("Nome...: ")

def pesquisa(nome):
    """
    Pesquisa na agenda um contato pelo nome, ignorando maiúsculas/minúsculas.
    Retorna o índice do contato ou None se não encontrado.
    """
    name = nome.lower()
    for i, contato in enumerate(agenda):
        if contato[0].lower() == name:
            return i
    return None

def editar():
    """
    Permite editar um contato existente, preservando o nome original.
    """
    pos = pesquisa(p_nome())
    if pos is not None:
        nome = agenda[pos][0]
        print(f"Nome...: {nome}")
        celular = input("Celular...: ")
        email = input("E-mail...: ")
        agenda[pos] = [nome, celular, email]
        print("\n-----------------------------")
        print("Contato editado com sucesso!")
        print(f"Nome: {nome} | Celular: {celular} | E-mail: {email}")
        print("-----------------------------\n")
    else:
        print("Contato não encontrado!\n")

def listar_dados(nome, celular, email):
    """
    Exibe os dados de um contato formatados.
    """
    print(f"Nome: {nome}\nCelular: {celular}\nE-mail: {email}\n")

def pesquisar():
    """
    Solicita o nome do contato e exibe seus dados se encontrado.
    """
    pos = pesquisa(p_nome())
    if pos is not None:
        nome, celular, email = agenda[pos]
        listar_dados(nome, celular, email)
    else:
        print("Contato não encontrado.\n")

def apagar():
    """
    Solicita o nome do contato e apaga-o da agenda após confirmação.
    """
    nome = p_nome()
    pos = pesquisa(nome)
    if pos is not None:
        nome_contato, celular, email = agenda[pos]
        print("Contato encontrado:")
        listar_dados(nome_contato, celular, email)
        confirm = input("Confirma a exclusão deste contato? [S/N]: ")
        if confirm.strip().lower() == 's':
            del agenda[pos]
            print("\n-----------------------------")
            print("Contato apagado com sucesso!")
            print("-----------------------------\n")
        else:
            print("Exclusão cancelada.\n")
    else:
        print("Contato não encontrado!\n")

def listar():
    """
    Exibe todos os contatos armazenados na agenda.
    """
    if not agenda:
        print("\nA agenda está vazia.\n")
        return

    print("\nCONTATOS DA AGENDA ######################\n")
    for contato in agenda:
        listar_dados(contato[0], contato[1], contato[2])
    print("FIM DA AGENDA ############################\n")

# Loop principal do programa
while True:
    opcao = menu()

    if opcao == 6:
        print("Encerrando o programa... Até logo!")
        break
    elif opcao == 1:
        novo()
    elif opcao == 2:
        editar()
    elif opcao == 3:
        pesquisar()
    elif opcao == 4:
        listar()
    elif opcao == 5:
        apagar()
    else:
        print("Opção inválida. Por favor, tente novamente.\n")

File: Projeto/Agenda.java
import java.util.ArrayList;
import java.util.Scanner;

public class Agenda {

    public static void main(String[] args) {

        Scanner entrada = new Scanner(System.in);
        ArrayList<String[]> agenda = new ArrayList<>();  // Lista que armazena os contatos (nome, celular, email)
        int opcao;
        String nome;

        do {
            exibeMenu();  // Exibe o menu principal

            System.out.print("Escolha uma opção: ");
            opcao = Integer.parseInt(entrada.nextLine());
            System.out.println();

            switch (opcao) {
                case 1:
                    // Adiciona um novo contato à agenda
                    novo(agenda, entrada);
                    break;

                case 2:
                    // Edita um contato existente, realizando busca por nome
                    System.out.println("_____ Editando (Pesquise o Contato):");
                    System.out.print("Digite o nome......: ");
                    nome = entrada.nextLine();
                    int linha = pesquisarContato(agenda, nome);
                    if (linha == -1) {
                        System.out.println("Contato não cadastrado!");
                    } else {
                        exibirContato(agenda, linha);
                        novo(agenda, entrada, linha); // Sobrescreve o contato encontrado
                    }
                    break;

                case 3:
                    // Pesquisa um contato pelo nome e exibe seus dados
                    System.out.println("_______ PESQUISAR CONTATO: ");
                    System.out.print("Digite o nome......: ");
                    nome = entrada.nextLine();
                    linha = pesquisarContato(agenda, nome);
                    if (linha == -1) {
                        System.out.println("Contato não cadastrado!");
                    } else {
                        exibirContato(agenda, linha);
                    }
                    break;

                case 4:
                    // Lista todos os contatos presentes na agenda
                    listarAgenda(agenda);
                    break;

                case 5:
                    // Apaga um contato pesquisado pelo nome, após confirmação
                    System.out.println("_______ EXCLUINDO (PESQUISE O CONTATO): ");
                    System.out.print("Digite o nome......: ");
                    nome = entrada.nextLine();
                    apagarContato(agenda, nome, entrada);
                    break;

                case 6:
                    // Finaliza a execução do programa
                    System.out.println("OBRIGADO POR UTILIZAR A NOSSA AGENDA :)");
                    break;

                default:
                    // Trata opções inválidas no menu
                    System.out.println("Opção inválida. Tente novamente.");
            }

            System.out.println();

        } while (opcao != 6);

        entrada.close();
    }

    /**
     * Remove um contato da agenda após confirmação do usuário.
     * @param agenda Lista de contatos.
     * @param nome Nome do contato a ser removido.
     * @param entrada Scanner para entrada de dados.
     */
    public static void apagarContato(ArrayList<String[]> agenda, String nome, Scanner entrada) {
        int linha = pesquisarContato(agenda, nome);
        if (linha != -1) {
            exibirContato(agenda, linha);
            System.out.print("Confirma a exclusão do contato? [S]im ou [N]ão: ");
            String opcao = entrada.nextLine();
            if (opcao.equalsIgnoreCase("s")) {
                agenda.remove(linha);
                System.out.println("Contato excluído com sucesso!");
            } else {
                System.out.println("Exclusão cancelada.");
            }
        } else {
            System.out.println("Contato não encontrado!");
        }
    }

    /**
     * Lista todos os contatos armazenados na agenda.
     * @param agenda Lista de contatos.
     */
    public static void listarAgenda(ArrayList<String[]> agenda) {
        System.out.println("_______ CONTATOS DA AGENDA: ");
        if (agenda.isEmpty()) {
            System.out.println("A agenda está vazia.");
        } else {
            for (int i = 0; i < agenda.size(); i++) {
                exibirContato(agenda, i);
                System.out.println("____________________________________");
            }
        }
        System.out.println("_______ FIM DA AGENDA.");
    }

    /**
     * Exibe os dados de um contato específico.
     * @param agenda Lista de contatos.
     * @param indice Índice do contato a ser exibido.
     */
    public static void exibirContato(ArrayList<String[]> agenda, int indice) {
        String[] contato = agenda.get(indice);
        System.out.println("Nome......: " + contato[0]);
        System.out.println("Celular...: " + contato[1]);
        System.out.println("E-mail....: " + contato[2]);
    }

    /**
     * Pesquisa um contato pelo nome na agenda.
     * @param agenda Lista de contatos.
     * @param nome Nome a ser pesquisado.
     * @return índice do contato encontrado ou -1 se não existir.
     */
    public static int pesquisarContato(ArrayList<String[]> agenda, String nome) {
        for (int i = 0; i < agenda.size(); i++) {
            if (agenda.get(i)[0].equalsIgnoreCase(nome)) {
                return i;
            }
        }
        return -1;
    }

    /**
     * Exibe o menu principal com as opções disponíveis.
     */
    public static void exibeMenu() {
        System.out.println("******* MENU *******");
        System.out.println("1 - Adicionar um novo contato");
        System.out.println("2 - Editar contato");
        System.out.println("3 - Pesquisar contato");
        System.out.println("4 - Listar contatos");
        System.out.println("5 - Apagar um contato");
        System.out.println("6 - Sair");
        System.out.println("********************");
    }

    /**
     * Adiciona um novo contato à agenda.
     * @param agenda Lista de contatos.
     * @param entrada Scanner para entrada de dados.
     */
    public static void novo(ArrayList<String[]> agenda, Scanner entrada) {
        System.out.println("______ PREENCHA O CONTATO: ");
        System.out.print("Nome......: ");
        String nome = entrada.nextLine();
        System.out.print("Celular...: ");
        String celular = entrada.nextLine();
        System.out.print("E-mail....: ");
        String email = entrada.nextLine();

        String[] contato = { nome, celular, email };
        agenda.add(contato);
        System.out.println("Contato adicionado com sucesso!");
        System.out.println("Contato: " + nome + " | " + celular + " | " + email);
    }

    /**
     * Atualiza os dados de um contato existente.
     * @param agenda Lista de contatos.
     * @param entrada Scanner para entrada de dados.
     * @param indice Índice do contato a ser atualizado.
     */
    public static void novo(ArrayList<String[]> agenda, Scanner entrada, int indice) {
        System.out.println("______ EDITANDO O CONTATO: ");
        System.out.print("Nome......: ");
        String nome = entrada.nextLine();
        System.out.print("Celular...: ");
        String celular = entrada.nextLine();
        System.out.print("E-mail....: ");
        String email = entrada.nextLine();

        String[] contato = { nome, celular, email };
        agenda.set(indice, contato);
        System.out.println("Contato editado com sucesso!");
        System.out.println("Contato atualizado: " + nome + " | " + celular + " | " + email);
    }
}
